### PR TITLE
Fixes #7979: Changing Directive version via API does not set default values, and return wrong parameters

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://www.rudder-project.org/redmine/issues/7979

### CODE REVIEW

**Comment by VinceMacBuche** [comment]
PR updated

File: rudder-core/src/main/scala/com/normation/rudder/domain/policies/Directive.scala [hunk 1]
```diff
@@ -196,7 +193,7 @@ object SectionVal {
         for {
           i <- 0 until cardinal
         } yield {
-          spec.getDirectVariables.map { vspec => (vspec.name, allValues(vspec.name)(i)) }.toMap
+          spec.getDirectVariables.map { vspec => (vspec.name, allValues.get(vspec.name).map(_(i)).getOrElse(vspec.constraint.default.getOrElse(""))) }.toMap
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/domain/policies/Directive.scala]: A comment will be useful for the future "us"

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala [hunk 2]
```diff
@@ -96,7 +96,7 @@ case class DirectiveAPIService2 (
     , technique       : Technique
     , activeTechnique : ActiveTechnique
     , directive       : Directive
-    , initialtState   : Option[Directive]
+    , initialState   : Option[Directive]
```
Comment (ncharles) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala]: wrong identation

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala [hunk 3]
```diff
@@ -319,56 +317,52 @@ case class DirectiveAPIService2 (
...
-        }
-
-      case eb:EmptyBox =>
-        val fail = eb ?~ (s"Could not find Directive ${directiveId.value}" )
+    ( for {
+      restDirective <- restValues ?~ (s"Could not extract values from request" )
+      // Find existing directive
+      (oldTechnique, activeTechnique, directive) <- readDirective.getDirectiveWithContext(directiveId) ?~ (s"Could not find Directive ${directiveId.value}" )
+
+      // Check if Technique version is chnaged (migration)
```
Comment (ncharles) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala]: typo - changed

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala [hunk 4]
```diff
@@ -319,56 +317,52 @@ case class DirectiveAPIService2 (
...
+      actualTechniqueId = TechniqueId(oldTechnique.id.name, restDirective.techniqueVersion.getOrElse(oldTechnique.id.version))
```
Comment (ncharles) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala]: actual is not the word you want.
Either oldTechniqueId / currentTechniqueId (if it's the one in place), or updatedTechniquedID (the new we want to set)
Actual means "real" in this context

Comment (ncharles) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala]: you mean updated in this case (even if the 'oldTechnique.id.name' got me there for a minute)

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala [hunk 5]
```diff
@@ -319,56 +317,52 @@ case class DirectiveAPIService2 (
...
+      actualTechnique <- (
+                            if (actualTechniqueId.version == oldTechnique.id.version) {
+                              Full(oldTechnique)
+                            } else {
+                              Box(techniqueRepository.get(actualTechniqueId))
+                            }
+                          ) ?~ s"Could not find technique ${actualTechniqueId.name} with version ${actualTechniqueId.version.toString}."
+
+       // Check parameters of the new Directive with the actuan technique version, IT will check that parameters are ok with the new technique
```
Comment (ncharles) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala]: actuan ?

**Comment by VinceMacBuche** [comment]
PR updated

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala (hunk 5 above)
Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/directive/DirectiveAPIService2.scala]: current :)

**Comment by VinceMacBuche** [comment]
PR updated

**Comment by VinceMacBuche** [comment]
PR updated

**Pull request merged**